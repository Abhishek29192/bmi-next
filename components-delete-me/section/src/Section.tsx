import React, { createContext, useContext } from "react";
import Typography, {
  Props as TypographyProps
} from "@bmi-digital/components/typography";
import Container, {
  Props as ContainerProps
} from "@bmi-digital/components/container";
import classnames from "classnames";
import styles from "./Section.module.scss";

export const SectionContext = createContext<boolean>(false);

export type BackgroundColor = "transparent" | "alabaster" | "white" | "pearl";

export type Props = {
  backgroundColor?: BackgroundColor;
  children: React.ReactNode;
  size?: "lg" | "md" | "sm" | "xl" | "xs" | false;
  spacing?: "default" | "none";
  className?: string;
  id?: string;
  hasNoPadding?: boolean;
  isSlim?: boolean;
  overflowVisible?: boolean;
} & ContainerProps;

const Section = ({
  backgroundColor = "transparent",
  children,
  size = "xl",
  spacing = "default",
  className,
  id,
  hasNoPadding = false,
  isSlim = false,
  overflowVisible = false,
  ...containerProps
}: Props) => {
  const isNested = useContext(SectionContext);

  if (isNested) {
    /* istanbul ignore next */
    if (process.env.NODE_ENV === "development") {
      // eslint-disable-next-line no-console
      console.warn(
        "Section: You cannot nest multiple section components - only children nodes will be rendered."
      );
    }

    return <>{children}</>;
  }

  return (
    <SectionContext.Provider value={true}>
      <div
        id={id}
        className={classnames(
          className,
          styles["Section"],
          backgroundColor !== "transparent" &&
            styles[`Section--${backgroundColor}`],
          spacing === "none" && styles["Section--no-spacing"],
          isSlim && styles["Section--slim"],
          overflowVisible && styles["Section--overflow-visible"]
        )}
      >
        <Container
          maxWidth={size}
          disableGutters={hasNoPadding}
          {...containerProps}
        >
          {children}
        </Container>
      </div>
    </SectionContext.Provider>
  );
};

const SectionTitle = ({ className, ...rest }: TypographyProps) => (
  <Typography
    variant="h2"
    hasUnderline
    className={classnames(styles["title"], className)}
    {...rest}
  />
);

Section.Title = SectionTitle;

export default Section;
