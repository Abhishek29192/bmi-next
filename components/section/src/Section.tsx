import React, { createContext, useContext } from "react";
import Typography, { Props as TypographyProps } from "@bmi/typography";
import Container from "@bmi/container";
import styles from "./Section.module.scss";
import classnames from "classnames";

export const SectionContext = createContext<boolean>(false);

export type Props = {
  backgroundColor?: "transparent" | "alabaster" | "white" | "pearl";
  children: React.ReactNode;
  size?: "lg" | "md" | "sm" | "xl" | "xs" | false;
  spacing?: "default" | "none";
};

const Section = ({
  backgroundColor = "transparent",
  children,
  size = "xl",
  spacing = "default"
}: Props) => {
  const isNested = useContext(SectionContext);

  if (isNested) {
    /* istanbul ignore next */
    if (process.env.NODE_ENV === "development") {
      // eslint-disable-next-line no-console
      console.warn(
        "Section: You cannot nest multiple section components - only children nodes will be rendered."
      );
    }

    return <>{children}</>;
  }

  return (
    <SectionContext.Provider value={true}>
      <div
        className={classnames(styles["Section"], {
          [styles[`Section--${backgroundColor}`]]:
            backgroundColor !== "transparent",
          [styles["Section--no-spacing"]]: spacing === "none"
        })}
      >
        <Container maxWidth={size}>{children}</Container>
      </div>
    </SectionContext.Provider>
  );
};

const SectionTitle = ({ className, ...rest }: TypographyProps) => (
  <Typography
    variant="h2"
    hasUnderline
    className={classnames(styles["title"], className)}
    {...rest}
  />
);

Section.Title = SectionTitle;

export default Section;
