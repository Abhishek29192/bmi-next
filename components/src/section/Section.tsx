import classnames from "classnames";
import React, { createContext, useContext } from "react";
import Container, { Props as ContainerProps } from "../container/Container";
import Typography, { Props as TypographyProps } from "../typography/Typography";
import { useStyles } from "./styles";

export const SectionContext = createContext<boolean>(false);

export type BackgroundColor = "transparent" | "alabaster" | "white" | "pearl";

export type Props = {
  backgroundColor?: BackgroundColor;
  children: React.ReactNode;
  size?: "lg" | "md" | "sm" | "xl" | "xs" | false;
  spacing?: "default" | "none";
  className?: string;
  id?: string;
  hasNoPadding?: boolean;
  isSlim?: boolean;
  overflowVisible?: boolean;
  isDialog?: boolean;
} & ContainerProps;

const Section = ({
  backgroundColor = "transparent",
  children,
  size = "xl",
  spacing = "default",
  className,
  id,
  hasNoPadding = false,
  isSlim = false,
  overflowVisible = false,
  isDialog = false,
  ...containerProps
}: Props) => {
  const isNested = useContext(SectionContext);
  const classes = useStyles();

  if (isNested) {
    /* istanbul ignore next */
    if (process.env.NODE_ENV === "development") {
      // eslint-disable-next-line no-console
      console.warn(
        "Section: You cannot nest multiple section components - only children nodes will be rendered."
      );
    }

    return <>{children}</>;
  }

  return (
    <SectionContext.Provider value={true}>
      <div
        id={id}
        className={classnames(
          className,
          !isDialog && classes.root,
          backgroundColor !== "transparent" &&
            ((backgroundColor === "white" && classes.white) ||
              (backgroundColor === "alabaster" && classes.alabaster) ||
              (backgroundColor === "pearl" && classes.pearl)),
          spacing === "none" && classes.noSpacing,
          isSlim && classes.slim,
          overflowVisible && classes.overflowVisible
        )}
      >
        <Container
          maxWidth={size}
          disableGutters={hasNoPadding}
          {...containerProps}
        >
          {children}
        </Container>
      </div>
    </SectionContext.Provider>
  );
};

const SectionTitle = ({ className, ...rest }: TypographyProps) => {
  const classes = useStyles();
  return (
    <Typography
      variant="h2"
      hasUnderline
      className={classnames(classes.title, className)}
      {...rest}
    />
  );
};

Section.Title = SectionTitle;

export default Section;
