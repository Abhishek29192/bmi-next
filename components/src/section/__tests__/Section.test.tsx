import React from "react";
import { renderWithThemeProvider } from "../../__tests__/helper";
import Section from "../Section";

describe("Section component", () => {
  it("renders correctly", () => {
    const { container } = renderWithThemeProvider(
      <Section>Hello world.</Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly a white section", () => {
    const { container } = renderWithThemeProvider(
      <Section backgroundColor="white">Hello world.</Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("does not render nested section components", () => {
    const { container } = renderWithThemeProvider(
      <Section>
        <Section>Hello world.</Section>
      </Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with a title", () => {
    const { container } = renderWithThemeProvider(
      <Section>
        <Section.Title>H2 Heading</Section.Title>
        Hello world.
      </Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with a custom className", () => {
    const { container } = renderWithThemeProvider(
      <Section className="test">
        <Section.Title>H2 Heading</Section.Title>
        Hello world.
      </Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with slim === true", () => {
    const { container } = renderWithThemeProvider(
      <Section isSlim>Hello world.</Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with overflow visible", () => {
    const { container } = renderWithThemeProvider(
      <Section overflowVisible>Hey, I like to over flow.</Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with spacing none", () => {
    const { container } = renderWithThemeProvider(
      <Section spacing="none">Hey, I like to over flow.</Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly as with size === 'lg'", () => {
    const { container } = renderWithThemeProvider(
      <Section size="lg">Hey, I like to over flow.</Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with id", () => {
    const { container } = renderWithThemeProvider(
      <Section id="id">Hey, I like to over flow.</Section>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with hasNoPadding", () => {
    const { container } = renderWithThemeProvider(
      <Section hasNoPadding={true}>Hey, I like to over flow.</Section>
    );
    expect(container).toMatchSnapshot();
  });
  it("renders correctly if rendered as dialog", () => {
    const { container } = renderWithThemeProvider(
      <Section isDialog>Not a section class applied</Section>
    );
    expect(container.querySelector(".Section")).toBeNull();
  });
});
