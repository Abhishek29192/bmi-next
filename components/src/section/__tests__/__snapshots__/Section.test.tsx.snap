// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section component does not render nested section components 1`] = `
<div>
  <div
    class="Section-root-21"
  >
    <div
      class="Container-root-29"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        Hello world.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly 1`] = `
<div>
  <div
    class="Section-root-1"
  >
    <div
      class="Container-root-9"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        Hello world.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly a white section 1`] = `
<div>
  <div
    class="Section-root-11 Section-white-15"
  >
    <div
      class="Container-root-19"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        Hello world.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly as with size === 'lg' 1`] = `
<div>
  <div
    class="Section-root-97"
  >
    <div
      class="Container-root-105"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        Hey, I like to over flow.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly with a custom className 1`] = `
<div>
  <div
    class="test Section-root-49"
  >
    <div
      class="Container-root-57"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        <h2
          class="MuiTypography-root Section-title-56 Typography-root-59 Typography-underline-61 MuiTypography-h2"
          data-testid="typography"
        >
          H2 Heading
        </h2>
        Hello world.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly with a title 1`] = `
<div>
  <div
    class="Section-root-31"
  >
    <div
      class="Container-root-39"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        <h2
          class="MuiTypography-root Section-title-38 Typography-root-41 Typography-underline-43 MuiTypography-h2"
          data-testid="typography"
        >
          H2 Heading
        </h2>
        Hello world.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly with hasNoPadding 1`] = `
<div>
  <div
    class="Section-root-117"
  >
    <div
      class="Container-root-125"
    >
      <div
        class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthXl"
      >
        Hey, I like to over flow.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly with id 1`] = `
<div>
  <div
    class="Section-root-107"
    id="id"
  >
    <div
      class="Container-root-115"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        Hey, I like to over flow.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly with overflow visible 1`] = `
<div>
  <div
    class="Section-root-77 Section-overflowVisible-78"
  >
    <div
      class="Container-root-85"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        Hey, I like to over flow.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly with slim === true 1`] = `
<div>
  <div
    class="Section-root-67 Section-slim-73"
  >
    <div
      class="Container-root-75"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        Hello world.
      </div>
    </div>
  </div>
</div>
`;

exports[`Section component renders correctly with spacing none 1`] = `
<div>
  <div
    class="Section-root-87 Section-noSpacing-92"
  >
    <div
      class="Container-root-95"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl"
      >
        Hey, I like to over flow.
      </div>
    </div>
  </div>
</div>
`;
